concurrency/MSNSession.java

```diff
@@ -0,0 +1,512 @@
+        Email email = Email.parseStr(getTransport().convertJIDToID(contact.getJID()));
+        if (email == null) {
+            Log.debug("MSN: Unable to remove illegal contact "+contact.getJID());
+            return;
+        }
+        msnMessenger.removeFriend(email, false);
+    }
+
+    /**
+     * @see net.sf.kraken.session.TransportSession#updateContact(net.sf.kraken.roster.TransportBuddy)
+     */
+    @Override
+    public void updateContact(MSNBuddy contact) {
+        Email email = Email.parseStr(getTransport().convertJIDToID(contact.getJID()));
+        if (email == null) {
+            Log.debug("MSN: Unable to update illegal contact "+contact.getJID());
+            return;
+        }
+        String nickname = getTransport().convertJIDToID(contact.getJID());
+        if (contact.getNickname() != null && !contact.getNickname().equals("")) {
+            nickname = contact.getNickname();
+        }
+
+        try {
+            MSNBuddy msnBuddy = getBuddyManager().getBuddy(contact.getJID());
+            if (msnBuddy.msnContact == null) {
+                MsnContact msnContact = msnMessenger.getContactList().getContactByEmail(email);
+                if (msnContact == null) {
+                    Log.debug("MSN: Contact updated but doesn't exist?  Adding.");
+                    addContact(contact.getJID(), nickname, (ArrayList<String>)contact.getGroups());
+                    return;
+                }
+                else {
+                    msnBuddy.setMsnContact(msnContact);
+                }
+            }
+            if (!msnBuddy.msnContact.getFriendlyName().equals(nickname)) {
+                msnMessenger.renameFriend(email, nickname);
+            }
+            syncContactGroups(email, (List<String>)contact.getGroups());
+        }
+        catch (NotFoundException e) {
+            Log.debug("MSN: Newly added buddy not found in buddy manager: "+email.getEmailAddress());
+        }
+    }
+    
+    /**
+     * @see net.sf.kraken.session.TransportSession#acceptAddContact(JID)
+     */
+    @Override
+    public void acceptAddContact(JID jid) {
+        final String userID = getTransport().convertJIDToID(jid);
+        Log.debug("MSN: accept-adding " + userID); 
+
+        // According to a packet dump made with Wireshark, 'accepting' a
+        // contact-add is done by adding the contact yourself (using an outgoing
+        // ADL).
+        final Email email = Email.parseStr(userID);
+        if (email == null) {
+            Log.warn("MSN: Unable to accept-add this illegal contact "+userID);
+            return;
+        }
+        
+        final TransportBuddyManager<MSNBuddy> manager = this.getBuddyManager();
+        
+        final String nickname;
+        if (manager.hasBuddy(jid)) {
+            try {
+                final MSNBuddy buddy = manager.getBuddy(jid);
+                nickname = buddy.getNickname();
+            } catch (NotFoundException ex) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)